flang/lib/Lower/Bridge.cpp

```diff
@@ -1965,62 +1965,87 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     Fortran::lower::BoxAnalyzer sba;
     sba.analyze(sym);
 
+    // compute extent from lower and upper bound.
+    auto computeExtent = [&](mlir::Value lb, mlir::Value ub) -> mlir::Value {
+      // let the folder deal with the common `ub - <const> + 1` case
+      auto diff = builder->create<mlir::SubIOp>(loc, idxTy, ub, lb);
+      auto one = builder->createIntegerConstant(loc, idxTy, 1);
+      return builder->create<mlir::AddIOp>(loc, idxTy, diff, one);
+    };
+
     // The origin must be \vec{1}.
-    auto populateShape = [&](auto &shapes, const auto &bounds) {
+    auto populateShape = [&](auto &shapes, const auto &bounds,
+                             mlir::Value box) {
+      int i = 0;
       for (auto *spec : bounds) {
-        if (auto low = spec->lbound().GetExplicit()) {
-          if (auto high = spec->ubound().GetExplicit()) {
-            Fortran::semantics::SomeExpr highEx{*high};
-            auto ub = createFIRExpr(loc, &highEx);
-            shapes.emplace_back(builder->createConvert(loc, idxTy, ub));
-          } else if (spec->ubound().isAssumed()) {
-            shapes.emplace_back(mlir::Value{});
-          } else {
-            TODO("upper bound");
-          }
+        assert(spec->lbound().GetExplicit() &&
+               "lbound must be explicit with constant value 1");
+        if (auto high = spec->ubound().GetExplicit()) {
+          Fortran::semantics::SomeExpr highEx{*high};
+          auto ub = createFIRExpr(loc, &highEx);
+          shapes.emplace_back(builder->createConvert(loc, idxTy, ub));
+        } else if (spec->ubound().isDeferred()) {
+          assert(box && "deferred bounds require a descriptor");
+          auto dim = builder->createIntegerConstant(loc, idxTy, i);
+          auto dimInfo = builder->create<fir::BoxDimsOp>(loc, idxTy, idxTy,
+                                                         idxTy, box, dim);
+          // TODO: are we sure the lower bound in the descriptor is also one in
+          // this case ? If not, we should use
```

Comment:
Right, and in the allocatable/pointers case, we would not be in the "all lower bounds are 1 case".
That means we need to add +1 to the descriptor upper bound here to get the extent.

If we ever start optimizing descriptor creation by reusing the ones we already have instead of creating some new at each call with assumed shape arguments, it may be risky to assume too much about the lower bound (One could argue that if the functions is not BIND(C), 18.5.3 does not apply since it is not meant to be called from C).
Let's assume that the lower bound is always zero in this case for now (I will need to update Calls.md design doc that is telling callee can assume the lower bounds is 1: https://github.com/flang-compiler/f18-llvm-project/blame/fir-dev/flang/docs/Calls.md#L441-L445).

Reply:
Never mind, the second field ofdescriptor is the extent, not the upper bound, so this comment is absolutely wrong.
Fixed that in the last commit.